File: Sources/PalmierPro/Models/TextAnimation.swift
import Foundation

struct WordTiming: Codable, Sendable, Equatable {
    var text: String
    var startFrame: Int
    var endFrame: Int
}

struct TextAnimation: Codable, Sendable, Equatable {
    var preset: Preset = .none
    var perWordFrames: Int = 6
    var highlight: TextStyle.RGBA?

    enum Preset: String, Codable, CaseIterable, Sendable {
        case none
        // Whole-clip / per-line.
        case fadeIn, popIn, slideUp, typewriter
        // Per word.
        case wordReveal, wordSlide, wordPop, wordCycle, highlightPop, highlightBlock

        enum RenderMode { case entrance, perWord, typewriter }

        var renderMode: RenderMode {
            switch self {
            case .none, .fadeIn, .popIn, .slideUp: .entrance
            case .typewriter: .typewriter
            case .wordReveal, .wordSlide, .wordPop, .wordCycle,
                 .highlightPop, .highlightBlock: .perWord
            }
        }

        var isPerWord: Bool { renderMode == .perWord }
        var usesHighlight: Bool { isPerWord }

        var displayName: String {
            switch self {
            case .none: L10n.key("Off")
            case .fadeIn: L10n.key("Fade In")
            case .popIn: L10n.key("Pop In")
            case .slideUp: L10n.key("Slide Up")
            case .typewriter: L10n.key("Typewriter")
            case .wordReveal: L10n.key("Word Reveal")
            case .wordSlide: L10n.key("Word Slide")
            case .wordPop: L10n.key("Word Pop")
            case .wordCycle: L10n.key("Word Cycle")
            case .highlightPop: L10n.key("Highlight")
            case .highlightBlock: L10n.key("Highlight Block")
            }
        }

        static let agentValues: [String] = ["off"] + allCases.filter { $0 != .none }.map(\.rawValue)

        static let perLine: [Preset] = [.fadeIn, .popIn, .slideUp, .typewriter]
        static let perWord: [Preset] = [.wordReveal, .wordSlide, .wordPop, .wordCycle,
                                        .highlightPop, .highlightBlock]
    }

    var isActive: Bool { preset != .none }

    static let defaultHighlight = TextStyle.RGBA(r: 1, g: 0.85, b: 0, a: 1)

    private enum CodingKeys: String, CodingKey { case preset, perWordFrames, highlight }

    init(preset: Preset = .none, perWordFrames: Int = 6, highlight: TextStyle.RGBA? = nil) {
        self.preset = preset
        self.perWordFrames = perWordFrames
        self.highlight = highlight
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.init(
            preset: (try? c.decode(Preset.self, forKey: .preset)) ?? .none,
            perWordFrames: (try? c.decode(Int.self, forKey: .perWordFrames)) ?? 6,
            highlight: try? c.decode(TextStyle.RGBA.self, forKey: .highlight)
        )
    }
}

File: windows/src/PalmierPro.Core/Models/TextAnimation.cs
namespace PalmierPro.Core.Models;

public record struct WordTiming(string Text, int StartFrame, int EndFrame);

public sealed class TextAnimation : IEquatable<TextAnimation>
{
    public TextAnimationPreset Preset { get; set; } = TextAnimationPreset.None;
    public int PerWordFrames { get; set; } = 6;
    public Rgba? Highlight { get; set; }

    public bool IsActive => Preset != TextAnimationPreset.None;

    public static readonly Rgba DefaultHighlight = new(1, 0.85, 0, 1);

    public bool Equals(TextAnimation? other)
        => other is not null && Preset == other.Preset && PerWordFrames == other.PerWordFrames && Highlight == other.Highlight;

    public override bool Equals(object? obj) => Equals(obj as TextAnimation);
    public override int GetHashCode() => HashCode.Combine(Preset, PerWordFrames, Highlight);
}

public enum TextAnimationPreset
{
    None,
    // Whole-clip / per-line.
    FadeIn,
    PopIn,
    SlideUp,
    Typewriter,
    // Per word.
    WordReveal,
    WordSlide,
    WordPop,
    WordCycle,
    HighlightPop,
    HighlightBlock,
}

public enum TextAnimationRenderMode
{
    Entrance,
    PerWord,
    Typewriter,
}

public static class TextAnimationPresetExtensions
{
    public static TextAnimationRenderMode RenderMode(this TextAnimationPreset preset) => preset switch
    {
        TextAnimationPreset.None or TextAnimationPreset.FadeIn or TextAnimationPreset.PopIn or TextAnimationPreset.SlideUp
            => TextAnimationRenderMode.Entrance,
        TextAnimationPreset.Typewriter => TextAnimationRenderMode.Typewriter,
        _ => TextAnimationRenderMode.PerWord,
    };

    public static bool IsPerWord(this TextAnimationPreset preset) => preset.RenderMode() == TextAnimationRenderMode.PerWord;
    public static bool UsesHighlight(this TextAnimationPreset preset) => preset.IsPerWord();
}

public enum TextFillMode
{
    Color,
    Footage,
}
